ksql-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryRoutingFunctionalTest.java

```diff
@@ -0,0 +1,388 @@
+      .withProperty(KsqlRestConfig.ADVERTISED_LISTENER_CONFIG, "http://localhost:8089")
+      .withProperty(KsqlRestConfig.KSQL_HEARTBEAT_ENABLE_CONFIG, true)
+      .withProperty(KsqlRestConfig.KSQL_HEARTBEAT_SEND_INTERVAL_MS_CONFIG, 600000)
+      .withProperty(KsqlRestConfig.KSQL_HEARTBEAT_CHECK_INTERVAL_MS_CONFIG, 200)
+      .withProperty(KsqlRestConfig.KSQL_HEARTBEAT_DISCOVER_CLUSTER_MS_CONFIG, 2000)
+      .withProperty(KsqlConfig.KSQL_QUERY_PULL_ENABLE_STALE_READS, true)
+      .withProperty(KsqlConfig.KSQL_STREAMS_PREFIX + "num.standby.replicas", 1)
+      .build();
+
+  private static final TestKsqlRestApp REST_APP_2 = TestKsqlRestApp
+      .builder(TEST_HARNESS::kafkaBootstrapServers)
+      .withProperty(KSQL_STREAMS_PREFIX + StreamsConfig.NUM_STREAM_THREADS_CONFIG, 1)
+      .withProperty(KSQL_STREAMS_PREFIX + StreamsConfig.STATE_DIR_CONFIG, getNewStateDir())
+      .withProperty(KsqlRestConfig.LISTENERS_CONFIG, "http://localhost:8087")
+      .withProperty(KsqlRestConfig.ADVERTISED_LISTENER_CONFIG, "http://localhost:8087")
+      .withProperty(KsqlRestConfig.KSQL_HEARTBEAT_ENABLE_CONFIG, true)
+      .withProperty(KsqlRestConfig.KSQL_HEARTBEAT_SEND_INTERVAL_MS_CONFIG, 600000)
+      .withProperty(KsqlRestConfig.KSQL_HEARTBEAT_CHECK_INTERVAL_MS_CONFIG, 200)
+      .withProperty(KsqlRestConfig.KSQL_HEARTBEAT_DISCOVER_CLUSTER_MS_CONFIG, 2000)
+      .withProperty(KsqlConfig.KSQL_QUERY_PULL_ENABLE_STALE_READS, true)
+      .withProperty(KsqlConfig.KSQL_STREAMS_PREFIX + "num.standby.replicas", 1)
+      .build();
+
+  @ClassRule
+  public static final RuleChain CHAIN = RuleChain
+      .outerRule(Retry.of(3, ZooKeeperClientException.class, 3, TimeUnit.SECONDS))
+      .around(TEST_HARNESS)
+      .around(REST_APP_0)
+      .around(REST_APP_1)
+      .around(REST_APP_2);
+
+  @Rule
+  public final ExpectedException expectedException = ExpectedException.none();
+
+  @BeforeClass
+  public static void setUpClass() {
+    //Create topic with 1 partition to control who is active and standby
+    TEST_HARNESS.ensureTopics(1, USER_TOPIC);
+
+    final AtomicLong timestampSupplier = new AtomicLong(BASE_TIME);
+
+    TEST_HARNESS.produceRows(
+        USER_TOPIC,
+        USER_PROVIDER,
+        VALUE_FORMAT,
+        timestampSupplier::getAndIncrement
+    );
+
+    //Create stream
+    makeAdminRequest(
+        "CREATE STREAM " + USERS_STREAM
+            + " (" + USER_PROVIDER.ksqlSchemaString() + ")"
+            + " WITH ("
+            + "   kafka_topic='" + USER_TOPIC + "', "
+            + "   key='" + USER_PROVIDER.key() + "', "
+            + "   value_format='" + VALUE_FORMAT + "'"
+            + ");"
+    );
+  }
+
+  @Before
+  public void setUp() {
+    REST_APP_0.start();
+    REST_APP_1.start();
+    REST_APP_2.start();
+    output = KsqlIdentifierTestUtil.uniqueIdentifierName();
+
+  }
+
+  @After
+  public void cleanUp() {
+    REST_APP_0.stop();
+    REST_APP_1.stop();
+    REST_APP_2.stop();
+  }
```

Comment:
why do you stop / start the servers on each test? This is going to make the tests run a _lot_ slower!

If you truly need them to start / stop on each test, then change them to be a `@Rule` rather than a `@ClassRule`.

Reply:
Done